### ISSUE TITLE
Add evaluation metrics for imbalanced data

### ISSUE THREAD

**Issue by Darkmyter** [scikit-multiflow/scikit-multiflow#73]
All the metrics available for classification evaluation are not useful when the dataset is imbalanced, especially the accuracy. An improvement could be made by adding metrics that reflect the performance of classifiers in such case.  The metrics could be F1-measure, G-mean...
This will be useful for evaluating ensemble methods and makes the evaluation easier when testing new algorithms instead of using scikit-learn.

**Comment by Darkmyter** [comment]
I will try to add these metrics!

**Comment by jacobmontiel** [comment]
Thanks @Darkmyter 

My suggestion is to start with F1 given its popularity. For this, it is necessary to add precision and recall to `ClassificationMeasurements`.

**Comment by Darkmyter** [comment]
Adding these metrics was direct thanks to the confusion matrix already present. Here is a test using HT on AGRAWAL (function 7).
![f1](https://user-images.githubusercontent.com/20753626/47640270-838b7480-db63-11e8-893f-3d77ce25b529.png)

The interruption observed in the current are because the window is too small to capture any ones. Unless the window is  very large it's not value to consider for evaluation i think .

**Comment by jacobmontiel** [comment]
Hello @Darkmyter, could you elaborate on why there are gaps in the plots?

**Comment by jacobmontiel** [comment]
How did you correct it? Could you add a code example?

What about balanced or semi-balanced cases?

**Comment by Darkmyter** [comment]
For base_evaluator I added the basic lines:
```            
if constants.F1_SCORE in self.metrics:
    print('{} - F1-score          : {:4f}'.format(
        self.model_names[i],
        self._data_buffer.get_data(metric_id=constants.F1_SCORE, data_id=constants.MEAN)[i]))
if constants.PRECISION in self.metrics:
    print('{} - Precision          : {:4f}'.format(
        self.model_names[i],
        self._data_buffer.get_data(metric_id=constants.PRECISION, data_id=constants.MEAN)[i]))
if constants.RECALL in self.metrics:
    print('{} - Recall          : {:4f}'.format(
        self.model_names[i],
        self._data_buffer.get_data(metric_id=constants.RECALL, data_id=constants.MEAN)[i]))
if constants.RUNNING_TIME in self.metrics:
```
```
elif metric == constants.F1_SCORE:
    for i in range(self.n_models):
        values[0].append(self.mean_eval_measurements[i].get_f1_score())
        values[1].append(self.current_eval_measurements[i].get_f1_score())
elif metric == constants.PRECISION:
    for i in range(self.n_models):
        values[0].append(self.mean_eval_measurements[i].get_precision())
        values[1].append(self.current_eval_measurements[i].get_precision())
elif metric == constants.RECALL:
    for i in range(self.n_models):
        values[0].append(self.mean_eval_measurements[i].get_recall())
        values[1].append(self.current_eval_measurements[i].get_recall())
```
And for mesasure_collection I added the functions to calculate the metrics:

```
    def get_precision(self):

        """ compute the precision of the classifier.

        Returns
        -------
        float
            The precision
        """
        if  self.confusion_matrix.value_at(1, 1)  + self.confusion_matrix.value_at(1, 0) == 0:
            return 0
        else:
            return self.confusion_matrix.value_at(1, 1) / \
                    ( self.confusion_matrix.value_at(1, 1)  + self.confusion_matrix.value_at(1, 0))

    def get_recall(self):

        """ Compute the recall.
        Returns
        -------
        float
            The recall.
        """
        if  self.confusion_matrix.value_at(1, 1)  + self.confusion_matrix.value_at(0, 1) == 0:
            return 0
        else:
            return self.confusion_matrix.value_at(1, 1) / \
                    ( self.confusion_matrix.value_at(1, 1)  + self.confusion_matrix.value_at(0, 1))

```
I also changed to confusion matrix to be created with size 2x2 in measure_collection `__init__` , because we need all the four elements of it the calculate the f1 and not wait till we get a one the increase it's size, this has no effect on other metrics as for any classification problem there are at least two classes. 

For the gaps i was initially trying to catch the zero division with `ZeroDivisionError` and that was working as it returns nan and `ZeroDivisionError` so changed it to simple checking with equal 0.

The metrics will work with any type of data whether balanced or not, but it's up to the user to choose the right window size that he sees for the evaluation.

**Comment by jacobmontiel** [comment]
OK, sounds good. It is important to specify in the documentation that this implementation only works for the binary case (taking 1 as the reference class). Later we can extend it to multiclass.

Another concern is to "hide" the handling of zero division from the user. We should raise a warning (once) when this happens. For example in `sklearn` you get this warning:

```
UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
```

Also, consider the following change to make the precision and recall calculations easier to read:

```python
def get_precision(self):

        """ compute the precision of the classifier.

        Returns
        -------
        float
            The precision
        """
        tp = self.confusion_matrix.value_at(1, 1)
        fp = self.confusion_matrix.value_at(1, 0)
        if tp + fp == 0:
            return 0.0
        else:
            return tp / ( tp + fp )

```
